Addition of documentation for secret rotations with FluxCD (#5)
stackhpc/capi-helm-fluxcd-config: docs/secret-rotation.md

> # Secret Rotations &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; **FluxCD**

Reviewer: The formatting looks odd in different browser window sizes, let's just keep things as simple as possible.
```suggestion
# Secret Rotation
```

This brief guide should outline all the steps needed to be able to swap out a set of secrets or credentials with FluxCD GitOps.

## [**Pre-requisites**](#pre-requisites-link)
#### [Useful Resources](#useful-resources-link)

> #### [Applications & Software](#applications--software-link)

Reviewer: Should these be inside the `Outline` section?

Author: I wouldn't say so, they aren't part of outlining the steps, or what is to come, and are things that are required before even starting.

> ## [**Outline**](#outline-link)

Reviewer: Let's clean up the formatting of this table of contents please. Use something more like [this](https://github.com/azimuth-cloud/capi-helm-charts/blob/main/charts/openstack-cluster/README.md#contents--) which you can see the raw markdown for [here](https://raw.githubusercontent.com/azimuth-cloud/capi-helm-charts/refs/heads/main/charts/openstack-cluster/README.md).

Author: Done

#### [Cluster Maintenance](#cluster-maintenance-link)

> &nbsp; &nbsp; &nbsp; &nbsp; - Basics of what is done in a standard secret rotation.

Reviewer: Please remove all instances of `&nbsp;`. Markdown should use tabs for indentation where necessary (and necessary usually only means to create sub-lists inside lists).

Author: Done

#### [Security Threat](#security-threat-link)
&nbsp; &nbsp; &nbsp; &nbsp; - What to do in the case of a compromised secret.

> ## [**Horizon Application Credentials**](#horizon-application-credentials-link)

Reviewer: Markdown headings (`##`) don't need more bold `**`. 
```suggestion
## [Horizon Application Credentials](#horizon-application-credentials-link)
```

#### [Create/Delete Credentials](#createdelete-credentials-link)
&nbsp; &nbsp; &nbsp; &nbsp; - How to create & delete application credentials within Horizon.

## [**Processing Secrets**](#processing-secrets-link)
#### [Encrypting Credentials](#encrypting-new-credential-link)
&nbsp; &nbsp; &nbsp; &nbsp; - Using Kubeseal to encrypt the newly created credentials.
#### [Updating Secrets](#updating-secrets-link)
&nbsp; &nbsp; &nbsp; &nbsp; - How to rotate new secrets using FluxCD’s GitOps.
#### [Deleting Old Secrets](#deleting-old-secrets-link)
&nbsp; &nbsp; &nbsp; &nbsp; - Some housekeeping steps post secret rotation.
#### [Rotating Compromised Secrets](#rotating-compromised-secrets-link)
&nbsp; &nbsp; &nbsp; &nbsp; - Summary for the rotation of secrets posing a security threat.

> ## [**Validate Configuration**](#validate-configuration-link)

Reviewer: ```suggestion
## [Validate Configuration](#validate-configuration-link)
```

#### [Sonobuoy](#sonobuoy-link)
&nbsp; &nbsp; &nbsp; &nbsp; - Brief outline on validating the Kubernetes configuration.

#
<a id="pre-requisites-link"></a>

> # **Pre-Requisites**

Reviewer: ```suggestion
# Pre-Requisites
```

This document assumes that a self-managed CAPI cluster deployed with FluxCD is already configured and deployed, if not, one can be deployed by following the instructions found in [this](https://github.com/stackhpc/capi-helm-fluxcd-config) GitHub repository.

In addition to this, the user trying to rotate the secrets should have an account on Horizon with the necessary permissions.

> <a id="useful-resources-link"></a>

Reviewer: HTML links aren't needed, you can refer to native markdown headings directly e.g. https://github.com/stackhpc/capi-helm-fluxcd-config/blob/secret-rotation-docs/docs/secret-rotation.md#useful-resources-1 (the `-1` suffix is only because you've used headings in your outline which should also be fixed by cleaning up the formatting as per my earlier comment).
```suggestion
```

> ### **Useful Resources**
> 
> [CAPI cluster FluxCD repository](https://github.com/stackhpc/capi-helm-fluxcd-config)
> 
> [Waldur FluxCD Demo repository](https://github.com/stackhpc/waldur-fluxcd-gitops-demo)
> 
> [CAPI Cluster debugging manual](https://github.com/azimuth-cloud/capi-helm-charts/blob/main/charts/openstack-cluster/DEBUGGING.md)

Reviewer: Let's remove this section please, we especially don't need to refer back to the repo we're already looking at. The debugging docs are indeed useful (and we should link to them in a separate `docs/debugging.md` page in this repo eventually, but they're not relevant to secret rotation).
```suggestion
```

> <a id="applications--software-link"></a>
> ### **Applications & Software**

Reviewer: Remove all links and extra `**` from headings please.
```suggestion
### Applications & Software
```

| This document assumes that the Kubernetes cluster’s ```kubeconfig``` is known. |

> | :---- |

Reviewer: Can we replace these types of box formatting with something which looks better in GitHub-rendered markdown like these please https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts

Reviewer: Also the `kubeconfig` should be in your cluster config repo at `clusters/<cluster-name/kubeconfig` so probably worth mentioning that.

Author: I love those markdown Alert boxes, they are great! Thanks, and I will add the bit about the directory in too.

Kubeseal

Current CAPI cluster FluxCD config

Horizon account

k9s (for debugging, optional)

<a id="outline-link"></a>
# **Outline**

The overall idea is to take advantage of Kubeseal’s encryption, allowing us to upload secrets to our repositories; meaning that after a branch containing the new secrets is merged into the branch being tracked by FluxCD, the FluxCD GitOps flow will update the current cluster deployment to reflect the newly merged branch.

Although the basic outline will remain the same, there are some variations in the preliminary actions depending on the situation and reason why secrets are being rotated.

The two main categories fall under **cluster maintenance** and **security threat**, and the differences in these situations will be further explained below.

<a id="cluster-maintenance-link"></a>
### **Cluster Maintenance**

It is often a good idea and good security practice to rotate passwords and secrets, as it reduces the chances of them existing long enough for them to be exploited or end up in the wrong hands; often many secrets will have an expiration date precisely for this reason.

> Therefore, in the case of the user simply updating secrets for the purpose of cluster maintenance it is simply enough to replace the current Kubeseal encrypted secret with an updated, manually encrypted, Kubeseal secret via a GitHub pull request and merge. Then delete the old application credential from Horizon.

Reviewer: Nit: Personally I'd format all references to the `kubeseal` tool as lower case inside backticks, as they do in the official [docs](https://github.com/bitnami-labs/sealed-secrets), but maybe that's just me...
```suggestion
Therefore, in the case of the user simply updating secrets for the purpose of cluster maintenance it is simply enough to replace the current Kubeseal encrypted secret with an updated, manually encrypted, `kubeseal` secret via a GitHub pull request and merge. Then delete the old application credential from Horizon.
```

Now a new file called **credentials.yaml** can be created in its place with the following structure:

```
apiVersion: v1
kind: Secret
metadata:
  name: <MAKE SURE THIS MATCHES “cloudCredentialsSecretName:” IN configmap.yaml>
  namespace: capi-self
  annotations:
    # Allow the sealed secret controller to take over this secret after bootstrapping
    sealedsecrets.bitnami.com/managed: "true"
stringData:
  clouds.yaml: |
    <INSERT THE CONTENTS OF YOUR NEW SECRET’S CLOUDS.YAML, PAYING ATTENTION TO THE
    INDENTATIONS>
```

Once the newly created **clouds.yaml**’s contents have been copied into **credentials.yaml**, it should look something like this:

```
apiVersion: v1
kind: Secret
metadata:
  name: <MAKE SURE THIS MATCHES “cloudCredentialsSecretName:” IN configmap.yaml>
  namespace: capi-self
  annotations:
    # Allow the sealed secret controller to take over this secret after bootstrapping
    sealedsecrets.bitnami.com/managed: "true"
stringData:
  clouds.yaml: |
    clouds:
      openstack:
        auth:
          auth_url: ...
          application_credential_id: ...
          application_credential_secret: ...
        interface: ...
        identity_api_version: 3

>         auth_type: ...

Reviewer: This is missing the `project_id` field as mentioned [here](https://github.com/stackhpc/capi-helm-fluxcd-config/tree/secret-rotation-docs?tab=readme-ov-file#usage) and explained in more detail [here](https://github.com/azimuth-cloud/capi-helm-charts/blob/main/charts/openstack-cluster/README.md#openstack-credentials). Not sure if / how you managed to test this secret rotation process without the `project_id` field in your new credentials? Unless the project ID requirement has been removed in the upstream `gophercloud` library and I've just never noticed 😄

Author: We discussed this previously

```

> From here make sure that the terminal’s current working directory is in the same one as **credentials.yaml**, `fluxcd-config/clusters/<cluster-name>/`, and then run the following Kubeseal command:

Reviewer: ```suggestion
From here make sure that the terminal’s current working directory is in the same one as **credentials.yaml**, `clusters/<cluster-name>/` from the repo root, and then run the following Kubeseal command:
```

### **Updating Secrets**

With the new secrets encrypted it is time to change the sealed secret which is on the kubernetes cluster.

<!-- The image on the right is the first sealed secret which is going to replaced. This is being viewed using the **k9s** UI. -->

The process of changing the secret over is as simple as adding, committing and pushing the new **credentials.yaml** to the upstream GitHub repository which is being tracked by FluxCD.

| It is highly recommended to `git switch -c` to a new branch when pushing the new credentials.yaml so that a pull request can be reviewed before merging to the tracked branch. |
| :---- |

When the new **credentials.yaml** has been merged into the FluxCD tracked repository branch, FluxCD will try to update the **sealed secret** on the self managed cluster once it recognises that there are differences between the cluster’s current configuration and the one on GitHub.

| The time interval in which FluxCD will check for these *drifts between configurations* is set in the cluster’s **helmrelease.yaml** found within the `fluxcd-config/components/<cluster-name>/` directory, under the `interval` variable. |
| :---- |

> This should result in the **sealed secret** on the kubernetes cluster to contain the new secret.

Reviewer: ```suggestion
This should result in the **sealed secret** on the Kubernetes cluster containing the new secret. The `sealed-secrets` controller on the cluster will then convert this to a standard k8s secret which can be [decoded](https://kubernetes.io/docs/tasks/configmap-secret/managing-secret-using-kubectl/#decoding-secret) using `kubectl` or `k9s` to check that the contents has been updated correctly.
```

<a id="deleting-old-secrets-link"></a>
###	**Deleting Old Secrets**

> Once the sealed secret has been updated on kubernetes it is time to do some *housekeeping* by now deleting the old credentials and unnecessary files created along the way, including, but not limited to, **old-credentials.yaml** and its corresponding application credential on Horizon.

Reviewer: Why 'not limited to'? Let's try to be precise about what does and doesn't need deleted please 🙂

Author: Reworded.

<a id="rotating-compromised-secrets-link"></a>
###	**Rotating Compromised Secrets**
As mentioned above, an application credential which is a security concern should be deleted before any other step; after which a new application credential, which will replace it, can be created.

| Don’t forget to download the clouds.yaml file after creating the new application credential\! |
| :---- |

Due to having deleted the old application credentials there is no need to rename the **credentials.yaml** file and can instead be directly replaced with the template, plus the contents of the new application credential’s **clouds.yaml**.

```
apiVersion: v1
kind: Secret
metadata:
  name: <MAKE SURE THIS MATCHES “cloudCredentialsSecretName:” IN configmap.yaml>
  namespace: capi-self
  annotations:
    # Allow the sealed secret controller to take over this secret after bootstrapping
    sealedsecrets.bitnami.com/managed: "true"
stringData:
  clouds.yaml: |
    clouds:
      openstack:
        auth:
          auth_url: ...
          application_credential_id: ...

>           application_credential_secret: ...

Reviewer: Missing `project_id` again here?

        interface: ...
        identity_api_version: 3
        auth_type: ...
```

Then, much like the previous secret rotation, use Kubeseal to encrypt the new secret. This time round, the Kubeseal command below will seal the secret and overwrite **credentials.yaml** rather than create an intermediate **encrypted-creds.yaml** file.

> | kubeseal \--kubeconfig kubeconfig \--format yaml \--controller-name sealed-secrets \--controller-namespace sealed-secrets-system \--secret-file credentials.yaml  \--sealed-secret-file credentials.yaml |
> | :---- |

Reviewer: Can we reformat code blocks / commands so that they're more readable please? E.g. wrap them in triple back ticks ``` and split them across multiple lines
```suggestion
kubeseal \
  --kubeconfig kubeconfig \
  --format yaml \
  --controller-name sealed-secrets \
  --controller-namespace sealed-secrets-system \
  --secret-file credentials.yaml  \
  --sealed-secret-file credentials.yaml
```
More details [here](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code). This also has the advantage that GitHub will render a little button on the page that lets you easily copy the code to your clipboard for copy/pasting.

Again, this file will then be added, committed and pushed to a preferably new, GitHub branch which can then be reviewed and merged into the branch which FluxCD is tracking for changes. After which, once per interval, FluxCD will compare the current configuration with the branch on GitHub, and upon noticing any differences will attempt to update the current cluster’s configuration to reflect the changes.

<a id="validate-configuration-link"></a>
# **Validate Configuration**

It is always a good idea to test your deployments for any potential risks or errors, which can lead to more issues further down the line. Therefore, it is worthwhile to validate the Kubernetes cluster with a diagnostic tool.

<a id="sonobuoy-link"></a>

> ###	**Sonobuoy**

Reviewer: Does the Sonobuoy output contain any confirmation that the new cloud credentials worked? E.g. by provisioning a `PVC` which creates a Cinder volume or a k8s `Service` of type `LoadBalancer` which triggers the creation of an OpenStack LB? This was the main motivation for including Sonobuoy in the secret rotation process.

Author: Discussed last week and how doesn't share if it was a success or not.
